core/src/main/scala/com/softwaremill/session/SessionManager.scala

```diff
@@ -84,7 +84,19 @@ trait CsrfManager[T] {
   def config: SessionConfig
 
   def tokenInvalidRejection = AuthorizationFailedRejection
-  def createToken(): String = SessionUtil.randomString(64)
+  def createToken(): String = {
+    val timestamp = System.currentTimeMillis().toString
+    val hmac = generateHmac(timestamp)
+    s"$hmac$timestamp"
+  }
+  def validateToken(token: String): Boolean = {
+    val len = System.currentTimeMillis().toString.length
+    token.nonEmpty && token.length > len && {
+      val (hmac, timestamp) = (token.dropRight(len), token.takeRight(len))
+      hmac == generateHmac(timestamp)
```

Comment:
maybe we shoudl use `Utils.constantTimeEquals` here?

Reply:
Done, thanks for pointing me to this function.